### PR TITLE
chore: cleanup mocha/chai/sinon usage across the repo

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Use consistent test runner. 

**Description**

- Cleanup mocha/chai usage across the repo 
- Split the test config files for unit/e2e/spec/browsers 
- Use consistent configuration for each package

**Steps to test or reproduce**

- Run all tests.

### CODE REVIEW

File: packages/beacon-node/test/unit/chain/validation/attestation/getShufflingForAttestationVerification.test.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
+import {afterEach, beforeEach, describe, expect, it, vi} from "vitest";
```
Comment (nazarhussain) [inline packages/beacon-node/test/unit/chain/validation/attestation/getShufflingForAttestationVerification.test.ts:1]: Spliced top level `describe` file from `attestation.test.ts` as the nature of tests requires different mocks.

File: packages/beacon-node/test/unit/chain/validation/attestation/validateAttestation.test.ts [hunk 2]
```diff
@@ -0,0 +1,316 @@
+import {BitArray} from "@chainsafe/ssz";
```
Comment (nazarhussain) [inline packages/beacon-node/test/unit/chain/validation/attestation/validateAttestation.test.ts:1]: Spliced top level `describe` file from `attestation.test.ts` as the nature of tests requires different mocks.

File: packages/beacon-node/test/unit/chain/validation/attestation/validateGossipAttestationsSameAttData.test.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
+import bls from "@chainsafe/bls";
```
Comment (nazarhussain) [inline packages/beacon-node/test/unit/chain/validation/attestation/validateGossipAttestationsSameAttData.test.ts:1]: Spliced top level `describe` file from `attestation.test.ts` as the nature of tests requires different mocks.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -39,15 +39,25 @@ Contributing to tests:
...
-- Spec tests often compare full expected vs actual states in JSON format. To better understand the diff it's convenient to use mocha's option `--inline-diffs`.
-- A single logical error can cause many spec tests to fail. To focus on a single test at a time you can use mocha's option `--bail` to stop at the first failed test
-- To then run only that failed test you can run against a specific file as use mocha's option `--grep` to run only one case
+- Spec tests often compare full expected vs actual states in JSON format.
+- A single logical error can cause many spec tests to fail. To focus on a single test at a time you can use vitest's option `--bail` to stop at the first failed test
+- To then run only that failed test you can run against a specific file as use vitest's filters to run only one case
 
 ```sh
-LODESTAR_PRESET=minimal ../../node_modules/.bin/mocha --config .mocharc.spec.yml test/spec/phase0/sanity.test.ts --inline-diffs --bail --grep "attestation"
+LODESTAR_PRESET=minimal ../../node_modules/.bin/vitest --config vitest.config.spec.ts --run test/spec/phase0/sanity.test.ts
```
Comment (nflaig) [inline CONTRIBUTING.md]: I think this command can be simplified, from previous discussion on this, I think we don't need to reference the config file and can use `yarn vitest` instead of pointing to the  executable

There is a related PR open with some discussion on it, see https://github.com/ChainSafe/lodestar/pull/6247#discussion_r1439943030
- https://github.com/ChainSafe/lodestar/pull/6247

File: packages/beacon-node/test/e2e/doppelganger/doppelganger.test.ts [hunk 5]
```diff
@@ -90,11 +91,12 @@ describe.skip("doppelganger / doppelganger test", function () {
     const pubKey = validatorUnderTest.validatorStore.votingPubkeys()[0];
     const beaconBlock = ssz.allForks.phase0.BeaconBlock.defaultValue();
 
+    // Signing should be possible if starting at genesis since doppelganger should be off
     await expect(
-      validatorUnderTest.validatorStore.signBlock(fromHexString(pubKey), beaconBlock, bn.chain.clock.currentSlot),
-      "Signing should be possible if starting at genesis since doppelganger should be off"
-    ).to.eventually.be.fulfilled;
+      validatorUnderTest.validatorStore.signBlock(fromHexString(pubKey), beaconBlock, bn.chain.clock.currentSlot)
+    ).resolves.toBeUndefined();
```
Comment (nflaig) [inline packages/beacon-node/test/e2e/doppelganger/doppelganger.test.ts]: if this assertion fails now the error message will be pretty meaningless, can we use withMessage here?

File: packages/beacon-node/test/e2e/network/reqresp.test.ts [hunk 6]
```diff
@@ -84,53 +84,6 @@ function runTests({useWorker}: {useWorker: boolean}): void {
     return [netA, netB, await getPeerIdOf(netA), await getPeerIdOf(netB)];
   }
 
-  // it("should send/receive a ping message", async function () {
```
Comment (nflaig) [inline packages/beacon-node/test/e2e/network/reqresp.test.ts:87]: Do you know why those were commented out in the first place and can be removed now?

File: packages/beacon-node/test/unit/network/peers/priorization.test.ts [hunk 7]
```diff
@@ -14,6 +14,7 @@ import {RequestedSubnet} from "../../../../src/network/peers/utils/index.js";
 
 type Result = ReturnType<typeof prioritizePeers>;
 
+// eslint-disable-next-line vitest/valid-describe-callback
 describe("network / peers / priorization", async () => {
```
Comment (nflaig) [inline packages/beacon-node/test/unit/network/peers/priorization.test.ts:18]: why is this eslint disable required?

File: .eslintrc.js [hunk 8]
```diff
@@ -201,24 +206,18 @@ module.exports = {
...
-        // It's common to call function inside describe block
-        // https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-setup-in-describe.md
-        "mocha/no-setup-in-describe": "off",
-        // We use to split before in small isolated tasks
-        // https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-sibling-hooks.md
-        "mocha/no-sibling-hooks": "off",
+        "vitest/consistent-test-it": ["error", {fn: "it", withinDescribe: "it"}],
+        // We use a lot dynamic assertions so tests may not have usage of expect
+        "vitest/expect-expect": "off",
+        "vitest/no-disabled-tests": "error",
```
Comment (matthewkeil) [inline .eslintrc.js]: Mocha had this as a warning to highlight skipped test.  Perhaps we should leave this as a warn and come back and fix the skipped tests in a separate PR.

File: package.json [hunk 9]
```diff
@@ -47,69 +47,45 @@
   "devDependencies": {
     "@chainsafe/eslint-plugin-node": "^11.2.3",
     "@dapplion/benchmark": "^0.2.4",
-    "@types/chai": "^4.3.6",
-    "@types/chai-as-promised": "^7.1.6",
-    "@types/mocha": "^10.0.1",
+    "@types/mocha": "^10.0.6",
```
Comment (matthewkeil) [inline package.json:50]: Interesting! Does vitest use the mocha types for compatibility or something?

File: packages/fork-choice/test/unit/protoArray/computeDeltas.test.ts [hunk 10]
```diff
@@ -132,52 +132,6 @@ describe("computeDeltas", () => {
     }
   });
 
-  /**
-   * Starting Aug 2023, this test case is not valid because when an attestation is added
-   * to forkchoice, the block should come first, i.e. nextIndex should be a number
-   */
```
Comment (nflaig) [inline packages/fork-choice/test/unit/protoArray/computeDeltas.test.ts:138]: maybe @tuyennhv can confirm that this is ok to be removed, I would assume it was just commented out instead of fully removed for a reason

File: packages/api/package.json [hunk 11]
```diff
@@ -65,7 +65,7 @@
     "lint:fix": "yarn run lint --fix",
     "pretest": "yarn run check-types",
     "test": "yarn test:unit && yarn test:e2e",
-    "test:unit": "vitest --run --dir test/unit/ --coverage",
```
Comment (matthewkeil) [inline packages/api/package.json:68]: I know you were pushing for coverage recently.  I noticed this was turned off everywhere.  What was the decision to do this driven by?

File: packages/api/test/perf/compileRouteUrlFormater.test.ts [hunk 12]
```diff
@@ -3,6 +3,7 @@ import {compileRouteUrlFormater} from "../../src/utils/urlFormat.js";
 /* eslint-disable no-console */
 
 describe("route parse", () => {
+  // eslint-disable-next-line vitest/no-disabled-tests
```
Comment (matthewkeil) [inline packages/api/test/perf/compileRouteUrlFormater.test.ts]: I dont think we should remove all the skipped test warning under this PR.  Feels like these should be addressed under separate PR cover so that it is clear why they all went away and so it easier to find when they get removed or turned back on

**Review by nflaig — COMMENTED** [review]
I only did a really high-level review, looks mostly correct, left few comments

Please before merging this, make sure it is rebased against unstable to avoid issues like lock file diffs on unstable branch

File: packages/beacon-node/test/e2e/network/network.test.ts [hunk 13]
```diff
@@ -98,7 +97,7 @@ function runTests({useWorker}: {useWorker: boolean}): void {
 
   // Current implementation of discv5 consumer doesn't allow to deterministically force a peer to be found
   // a random find node lookup can yield no results if there are too few peers in the DHT
-  // it.skip("should connect to new peer by subnet", async function () {});
+  it.todo("should connect to new peer by subnet", async function () {});
```
Comment (matthewkeil) [inline packages/beacon-node/test/e2e/network/network.test.ts:100]: This is a sweet feature!!

**Review by matthewkeil — DISMISSED** [review]
Overall seems like a great update!!! 🎸  There are a few high level things that I think should be changed by this PR in the comments.  

There are a lot of code changes here and it will be easier to review a bit more closely on a call or something.  In particular how all the mocking was changed.

File: CONTRIBUTING.md (hunk 4 above)
Comment (matthewkeil) [reply CONTRIBUTING.md]: agreed

File: package.json (hunk 9 above)
Comment (nazarhussain) [reply package.json:50]: Our performance tests still use the mocha internally, so why we need those types.

File: .eslintrc.js (hunk 8 above)
Comment (nazarhussain) [reply .eslintrc.js]: I wanted no more `skip` tests introduced further and warnings are pretty much ignored by most of us (including myself). As discussed will remove from this PR for now.

File: packages/api/package.json (hunk 11 above)
Comment (nazarhussain) [reply packages/api/package.json:68]: It's just moved to the config file, so so we can control it via `process.env.CI` variable. This helps to run tests faster locally for quick developer feedback.

File: packages/beacon-node/test/unit/network/peers/priorization.test.ts (hunk 7 above)
Comment (nazarhussain) [reply packages/beacon-node/test/unit/network/peers/priorization.test.ts:18]: `vitest/recommended` eslint rules suggest to have no `async` for `describe` blocks and move such code into hooks.

File: packages/beacon-node/test/e2e/doppelganger/doppelganger.test.ts (hunk 5 above)
Comment (nazarhussain) [reply packages/beacon-node/test/e2e/doppelganger/doppelganger.test.ts]: Yes we can that was making the assertion looks odd. Updated it with `toBeWithMessage` now.

File: CONTRIBUTING.md (hunk 4 above)
Comment (nazarhussain) [reply CONTRIBUTING.md]: We can skip the config file for default config file path. But here we have custom config file so we have to pass it. Yes one can do `yarn vitest` instead of `../../node_modules/.bin/vitest`.

File: packages/beacon-node/test/e2e/network/reqresp.test.ts (hunk 6 above)
Comment (nazarhussain) [reply packages/beacon-node/test/e2e/network/reqresp.test.ts:87]: This was commented a year earlier in following PR. https://github.com/ChainSafe/lodestar/pull/5229/files#diff-ed1686203e4792b95a12ac893ea47ec4435a2c453ff81ccd14b0fa83943a82cc

I believe testing those messages are not necessary as we are testing connection and other messages already.

File: packages/beacon-node/test/unit/network/peers/priorization.test.ts (hunk 7 above)
Comment (matthewkeil) [reply packages/beacon-node/test/unit/network/peers/priorization.test.ts:18]: Silly question, do you think we can/should put the setup in a hook?  Seems like it can but am guessing you have a reason to go this route instead.

File: packages/reqresp/test/utils/index.ts [hunk 14]
```diff
@@ -21,14 +21,24 @@ export async function* arrToSource<T>(arr: T[]): AsyncGenerator<T> {
...
+  if (message) {
+    expect(chunks.map(toHexString).join("").replace(/0x/g, "")).toEqualWithMessage(
+      expectedChunks.map(toHexString).join("").replace(/0x/g, ""),
+      message
+    );
+  } else {
+    expect(chunks.map(toHexString).join("").replace(/0x/g, "")).toEqual(
+      expectedChunks.map(toHexString).join("").replace(/0x/g, "")
+    );
+  }
```
Comment (matthewkeil) [inline packages/reqresp/test/utils/index.ts:33]: Its interesting that this test case, and the one below, changed

File: packages/validator/test/unit/utils/clock.test.ts [hunk 15]
```diff
@@ -89,12 +89,11 @@ describe("util / Clock", function () {
...
     ];
-    for (const {name, delta} of testCase) {
-      it(name, async function () {
-        const currentSlot = getCurrentSlotAround(testConfig, genesisTime);
-        vi.advanceTimersByTime(delta * 1000);
-        expect(getCurrentSlotAround(testConfig, genesisTime)).toBe(currentSlot + 1);
-      });
-    }
+
+    it.each(testCase)("$name", async function ({delta}) {
```
Comment (matthewkeil) [inline packages/validator/test/unit/utils/clock.test.ts:93]: This syntax is so much better.  I really like vitest syntax.  Especially the mocks/stubs!!! 🌟

File: tsconfig.build.json [hunk 16]
```diff
@@ -25,10 +25,8 @@
     "incremental": true,
     "preserveWatchOutput": true,
 
-
-    // There are two duplicate type definitions included from `chai` and `vitest` packages.
-    // There is one invalid type declaration introduced from `webdriverio -> got` package. 
-    // TODO: Once we completely remove `chai` and upgrade `webdriverio` we can enable this check again.  
+    // TODO: Investigate following errors:
+    // - Cannot find module 'rollup/parseAst' or its corresponding type declarations
```
Comment (matthewkeil) [inline tsconfig.build.json:29]: 🤔

File: vitest.base.unit.config.ts [hunk 17]
```diff
@@ -17,6 +17,7 @@ export default defineConfig({
     setupFiles: [path.join(__dirname, "./scripts/vitest/customMatchers.ts")],
     reporters: ["default", "hanging-process"],
     coverage: {
+      enabled: process.env.CI === "true",
```
Comment (matthewkeil) [inline vitest.base.unit.config.ts:20]: Ahhh!!! I see why you removed the `--coverage` everywhere now

File: karma.base.config.js
Comment (matthewkeil) [inline karma.base.config.js:1]: Does karma go away completely an vitest does this now?  Guessing that is the case but figured it was worth asking explicitly

File: CONTRIBUTING.md (hunk 4 above)
Comment (matthewkeil) [inline CONTRIBUTING.md]: What do you think about this format?

```suggestion
LODESTAR_PRESET=minimal yarn vitest --run --config vitest.config.spec.ts test/spec/phase0/sanity.test.ts
```

**Review by matthewkeil — COMMENTED** [review]
Looks great overall!!  Thanks for taking this across the finish line.  There is a lot of great, detailed work here.  Just a couple of small questions that came up when i went through with a fine toothed comb.

File: CONTRIBUTING.md (hunk 4 above)
Comment (matthewkeil) [reply CONTRIBUTING.md]: https://github.com/ChainSafe/lodestar/pull/6311#discussion_r1459701393

File: packages/reqresp/test/utils/index.ts (hunk 14 above)
Comment (nazarhussain) [reply packages/reqresp/test/utils/index.ts:33]: It's just a condition to choose between toEqual and toEqualWithMessage. We can't pass undefined as message to the later assertion.

File: tsconfig.build.json (hunk 16 above)
Comment (nazarhussain) [reply tsconfig.build.json:29]: We had this ts config earlier for the packages and reason listed in comments. I tried to remove and verify, so with new update packages the old error is no more there but a new one pop up. 

Investigating that would be a different separate task.

File: karma.base.config.js
Comment (nazarhussain) [reply karma.base.config.js:1]: Yes mocha, chai, sinon, karma, nyc, c8 and few other dependencies goes away and replaced with vitest. 

Only exception for now is the performance tests, the benchmark package we use internally depend on mocha. I hope to contribute to that package to make it runner agnostic in near future.

File: packages/beacon-node/test/unit/network/peers/priorization.test.ts (hunk 7 above)
Comment (nazarhussain) [reply packages/beacon-node/test/unit/network/peers/priorization.test.ts:18]: There are multiple ways 

1. Do setup in test level hooks
2. Create a file for config level `setupFile` that execute before every test context. 
3. Create a file for config level `globalSetup` that execute before a test run. 

Te basic idea is any `async` code usually evolve some resources, so calling it in `beforeAll` or `beforeEach` and then clearing resources in `afterAll` or `afterEach` is preferred. If you do on describe level, there is more chances you forget to clear the resources.

File: tsconfig.build.json (hunk 16 above)
Comment (wemeetagain) [reply tsconfig.build.json:29]: pls create issues for this

**Review by wemeetagain — APPROVED** [review]

**Comment by nazarhussain** [comment]
@matthewkeil I am merging this PR to move further with remaining action points. But would be happy to answer any questions or work on feedback in upcoming PRs.

**Comment by matthewkeil** [comment]
> @matthewkeil I am merging this PR to move further with remaining action points. But would be happy to answer any questions or work on feedback in upcoming PRs. 

Looks good. No worries, was mostly just concerned about the other reviews and @wemeetagain gave a final blessing so all is gravy baby 😊. Good work!

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.15.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.15.0) :tada:
